### PR TITLE
bitflyer.fetchMarkets

### PR DESCRIPTION
**Pull request by samgermain**
Based on what I read at these links, it looks like FX is swap and the settle currency is always `JPY`

https://lightning.bitflyer.com/About-Fx
https://bitflyer.com/en-jp/commission

----------------------------

# Before

```
2022-02-08T04:59:07.077Z
Node.js: v14.17.0
CCXT v1.72.37
bitflyer.fetchMarkets ()
4113 ms
             id |          symbol | base | quote | baseId | quoteId | maker | taker |   type |  spot | future
-------------------------------------------------------------------------------------------------------------
        BTC_JPY |         BTC/JPY |  BTC |   JPY |    BTC |     JPY | 0.002 | 0.002 |   spot |  true |  false
...
     FX_BTC_JPY |      FX_BTC_JPY |  BTC |   JPY |    BTC |     JPY |     0 |     0 | future | false |   true
BTCJPY11FEB2022 | BTCJPY11FEB2022 |  BTC |   JPY |    BTC |     JPY |     0 |     0 | future | false |   true
BTCJPY18FEB2022 | BTCJPY18FEB2022 |  BTC |   JPY |    BTC |     JPY |     0 |     0 | future | false |   true
BTCJPY25MAR2022 | BTCJPY25MAR2022 |  BTC |   JPY |    BTC |     JPY |     0 |     0 | future | false |   true
        BTC_EUR |         BTC/EUR |  BTC |   EUR |    BTC |     EUR | 0.002 | 0.002 |   spot |  true |  false
        BTC_JPY |         BTC/JPY |  BTC |   JPY |    BTC |     JPY | 0.002 | 0.002 |   spot |  true |  false
15 objects
```

# After

```
2022-02-08T06:02:03.422Z
Node.js: v14.17.0
CCXT v1.72.37
bitflyer.fetchMarkets ()
4178 ms
             id |             symbol | base | quote | settle | baseId | quoteId | settleId |   type |  spot | margin |  swap | future | option | active | contract | linear | inverse | taker | maker | contractSize |        expiry |           expiryDatetime | strike | optionType | precision |                                           limits
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
        BTC_JPY |            BTC/JPY |  BTC |   JPY |        |    BTC |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        XRP_JPY |            XRP/JPY |  XRP |   JPY |        |    XRP |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        ETH_JPY |            ETH/JPY |  ETH |   JPY |        |    ETH |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        XLM_JPY |            XLM/JPY |  XLM |   JPY |        |    XLM |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
       MONA_JPY |           MONA/JPY | MONA |   JPY |        |   MONA |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        ETH_BTC |            ETH/BTC |  ETH |   BTC |        |    ETH |     BTC |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        BCH_BTC |            BCH/BTC |  BCH |   BTC |        |    BCH |     BTC |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
     FX_BTC_JPY |        BTC/JPY:JPY |  BTC |   JPY |    JPY |    BTC |     JPY |          |   swap | false |  false |  true |  false |  false |   true |     true |   true |   false |     0 |     0 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
BTCJPY11FEB2022 | BTC/JPY:JPY-220211 |  BTC |   JPY |    JPY |    BTC |     JPY |          | future | false |  false | false |   true |  false |   true |     true |   true |   false |     0 |     0 |              | 1644559200000 | 2022-02-11T06:00:00.000Z |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
BTCJPY18FEB2022 | BTC/JPY:JPY-220218 |  BTC |   JPY |    JPY |    BTC |     JPY |          | future | false |  false | false |   true |  false |   true |     true |   true |   false |     0 |     0 |              | 1645164000000 | 2022-02-18T06:00:00.000Z |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
BTCJPY25MAR2022 | BTC/JPY:JPY-220325 |  BTC |   JPY |    JPY |    BTC |     JPY |          | future | false |  false | false |   true |  false |   true |     true |   true |   false |     0 |     0 |              | 1648188000000 | 2022-03-25T06:00:00.000Z |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        BTC_USD |            BTC/USD |  BTC |   USD |        |    BTC |     USD |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        BTC_JPY |            BTC/JPY |  BTC |   JPY |        |    BTC |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        BTC_EUR |            BTC/EUR |  BTC |   EUR |        |    BTC |     EUR |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
        BTC_JPY |            BTC/JPY |  BTC |   JPY |        |    BTC |     JPY |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         | 0.002 | 0.002 |              |               |                          |        |            |        {} | {"leverage":{},"amount":{},"price":{},"cost":{}}
15 objects
```

### CODE REVIEW

File: js/bitflyer.js [hunk 1]
```diff
@@ -101,58 +101,133 @@ module.exports = class bitflyer extends Exchange {
...
+                type = 'swap';
+                baseId = currencies[1]
+                quoteId = currencies[2]
+            } else if (future) {
+                const alias = this.safeString(market, 'alias');
+                const currencyId = this.safeString (alias.split('_'), 0);
+                baseId = currencyId.slice (0,-3);
+                quoteId = currencyId.slice (-3);
+                const date = this.safeString (id.split(currencyId), 1);
+                const rfc2616 = this.rfc2616(date);
```
Comment (kroitor) [inline js/bitflyer.js:164]: We don't have an RFC 2616 parser yet. It also has a few linting issues (missing spaces mostly).

File: js/bitflyer.js [hunk 2]
```diff
@@ -101,58 +101,133 @@ module.exports = class bitflyer extends Exchange {
...
+            if (swap) {
+                type = 'swap';
+                baseId = currencies[1]
+                quoteId = currencies[2]
+            } else if (future) {
+                const alias = this.safeString(market, 'alias');
+                const currencyId = this.safeString (alias.split('_'), 0);
+                baseId = currencyId.slice (0,-3);
+                quoteId = currencyId.slice (-3);
+                const date = this.safeString (id.split(currencyId), 1);
```
Comment (kroitor) [inline js/bitflyer.js:163]: Let's do the splits on separate lines to keep the code cleaner.

File: js/bitflyer.js (hunk 1 above)
Comment (samgermain) [reply js/bitflyer.js:164]: Is there a method that will work that isn't this?

Comment (kroitor) [reply js/bitflyer.js:164]: The base `rfc2616` does not really parse anything, it takes a timestamp in milliseconds and converts it to a rfc2616 string. We can probably use parseDate on it, or we could break down the date in parts.

Comment (samgermain) [reply js/bitflyer.js:164]: `parseDate` didn't work, I looked through the time functions and this was the only one that worked

Comment (samgermain) [reply js/bitflyer.js:164]: The date variable has the format `11FEB2022`

Comment (kroitor) [reply js/bitflyer.js:166]: I'll try to resolve that.

**Pull request merged**